modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp

```diff
@@ -0,0 +1,477 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#ifdef HAVE_ONEVPL
+
+#include <algorithm>
+#include <exception>
+
+#include <opencv2/gapi/streaming/onevpl/data_provider_interface.hpp>
+#include "streaming/onevpl/data_provider_defines.hpp"
+
+#include "streaming/onevpl/engine/transcode/transcode_engine_legacy.hpp"
+#include "streaming/onevpl/engine/transcode/transcode_session.hpp"
+#include "streaming/onevpl/accelerators/accel_policy_interface.hpp"
+#include "streaming/onevpl/accelerators/surface/surface.hpp"
+#include "streaming/onevpl/cfg_params_parser.hpp"
+#include "streaming/onevpl/utils.hpp"
+#include "logger.hpp"
+
+#define ALIGN16(value)           (((value + 15) >> 4) << 4)
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace onevpl {
+
+template<typename Type>
+bool set_vpp_param(const char* name, Type& out_vpp_param,
+                   const std::map<std::string, mfxVariant> &params_storage,
+                   mfxSession session);
+
+template<>
+bool set_vpp_param<uint32_t>(const char* name, uint32_t& out_vpp_param,
+                             const std::map<std::string, mfxVariant> &params_storage,
+                             mfxSession session) {
+    auto it = params_storage.find(name);
+    if (it != params_storage.end()) {
+        auto value = it->second.Data.U32;
+        GAPI_LOG_INFO(nullptr, "[" << session << "] set \"" << name <<
+                               "\": " << value);
+        out_vpp_param = value;
+        return true;
+    }
+    return false;
+}
+
+template<>
+bool set_vpp_param<uint16_t>(const char* name, uint16_t& out_vpp_param,
+                             const std::map<std::string, mfxVariant> &params_storage,
+                             mfxSession session) {
+    auto it = params_storage.find(name);
+    if (it != params_storage.end()) {
+        auto value = it->second.Data.U16;
+        GAPI_LOG_INFO(nullptr, "[" << session << "] set \"" << name <<
+                               "\": " << value);
+        out_vpp_param = value;
+        return true;
+    }
+    return false;
+}
```

Comment:
`Type` deducted from string `param_name` - so there would be concretization in caller code anyway like switch-case. 
By the way - i cannot see an approach how it may be written shorter ( and faster :))

Reply:
No, I meant
```
template<typename Type> Type get_vpp_param_value(MapIterator& it);

template<> uint16_t get_vpp_param_value<uint16_t>(MapIterator& it) {
    return it->second.Data.U16;
}

template<> uint32_t get_vpp_param_value<uint32_t>(MapIterator& it) {
    return it->second.Data.U32;
}

template<typename Type>
bool set_vpp_param(const char* name, Type& out_vpp_param,
                   const std::map<std::string, mfxVariant> &params_storage,
                   mfxSession session) {
    auto it = params_storage.find(name);
    if (it != params_storage.end()) {
        auto value = get_vpp_param_value<Type>(it);
        GAPI_LOG_INFO(nullptr, "[" << session << "] set \"" << name <<
                               "\": " << value);
        out_vpp_param = value;
        return true;
    }
    return false;
}
```

I believe <Type> can do its work, without switch-case

Reply:
great, thanks